### ISSUE TITLE
Improving Travis

### ISSUE THREAD

**Issue by RichJeanes** [2factorauth/twofactorauth#1265]
So I don't know how Rake works and how it's used to do the link checking, but I just wrote a one line bash command to get all of the URLs out of a commit. Think we can put it to use and stop checking every link every time?

``` bash
git show <commit> | grep ^+[[:space:]] | grep http | cut -c13-
```

`git show <commit>` - prints the diff
`grep ^+[[:space:]]` - finds any lines that begin with a plus sign followed by a space, meaning lines that the diff lists as new
`grep http` - grabs only the new lines that contain URLs
`cut -c13-` - removes the first twelve characters, which would be either "+      doc: " or "+      url: "

I was thinking we wrap it up as a bash script with error handling and logic to just produce a list of URLs. Then maybe just do a response check for 404's with wget or something?
We would need to think of a way to check every commit in a PR. Start at HEAD then check every commit back through the chain (just using the parent hash) until we hit a merge commit? Would that work?

**Comment by mxxcon** [comment]
From what I understand, it verifies urls here https://github.com/jdavis/twofactorauth/blob/master/Rakefile#L9
Looking at that file I noticed line 13. From what I understand(and I don't know ruby), it seems to be running checks in 8 threads. Maybe we can try to increase this number to parallelize these checks?
Also https://github.com/grosser/parallel#usage says that it supports `in_processes` or `in_threads`. Maybe `in_processes` will be faster as well?

**Comment by RichJeanes** [comment]
More threads won't necessarily make it faster if Travis doesn't give the VM more cores and [here](http://docs.travis-ci.com/user/ci-environment/) it says you only get 2 cores (burstable). Sure, it's waiting on blocking calls while waiting for the web server, but you're going to hit a ceiling at some point; that still isn't a scalable solution, it's just trying to throw more horsepower at it.

So instead of using HTMLProofer, we could use something like `sh 'check_links.sh'`, yeah?

Also, out of curiosity, it says on that Build Environment page that bundler and rake are already installed, so why does `bundle install` have to be called?

**Comment by mxxcon** [comment]
We don't need more cores for cpu intensive processing, but rather to launch as many checks in parallel as possible.
With close to 1000 links to check, each slow URL will delay everything else behind it. If instead of 8 we do 16, we can essentially half the time it takes to check these urls.

I don't know what kind of info is available to Travis at runtime. How would you pass `<commit>` value?

**Comment by RichJeanes** [comment]
But that's assuming that we're already waiting on at least 7 of those 8 threads. We might see improvements, but nowhere near as much as only checking new links, obviously.

You don't need to. The Travis build just clones the repo and, with no arguments, `show` operates on HEAD  from which you can extract the diff and the parent. You check the links in the diff, if any, then loop using the parent commit as the argument for `show`. If it's a merge commit, exit loop, else, check for links.

**Comment by smholloway** [comment]
Let me see if I understand... We have an issue with Travis timing out, which can be explained by the link-checker taking too long. The link-checker iterates over every link (and we have 1,000+), so every link is checked on every commit. Rich's change will allow us to check only the _new_ links for that commit. Sounds like a good thing to me.

**Comment by jamcat22** [comment]
The only problem is when old links stop working. Like with Charles Schwab.

**Comment by Carlgo11** [comment]
I suggest we have 2 scripts then. One for Travis and one for verifying locally.

**Comment by Carlgo11** [comment]
Should we disable html-proofer for the time being?

**Comment by jamcat22** [comment]
Yep.

**Comment by Carlgo11** [comment]
So @RichJeanes, could you create a PR? You can have the branch on the main repo if you want. That way we assist if you need any help.

**Comment by RichJeanes** [comment]
Is this still relevant? Apparently I never got around to implementing my idea xD

@stephengroat and @psgs seem to handle Travis now and I never really fully understood how the build process works anyway (mostly just from not taking the time to go through it and play around on my own).

**Comment by fpigerre** [comment]
I think @stephengroat fixed this issue in pull request #2478 with the Rake `proof_external` task.

**Comment by Carlgo11** [comment]
No longer applies to the project structure.
